evap/results/tools.py

```diff
@@ -171,42 +199,82 @@ def _calculate_results_impl(course):
+            non_grade_questions_distribution = normalize_distribution(non_grade_questions_counts)
+            contributor_distributions.append(mix_distributions(grade_questions_distribution, non_grade_questions_distribution, settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE))
+        else:
+            # calculate the grade question average, non-grade question average and the distributions for the course
+            course_grade_questions_average = grade_questions_average
+            course_grade_questions_distribution = normalize_distribution(grade_questions_counts)
+            course_non_grade_questions_average = non_grade_questions_average
+            course_non_grade_questions_distribution = normalize_distribution(non_grade_questions_counts)
+
+    # calculate the average grade and distribution over all contributors
+    contributors_average = avg(contributor_averages)
+    contributors_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+    for distribution in contributor_distributions:
+        for answer, value in distribution.items():
+            contributors_distribution[answer] += value
+    contributors_distribution = normalize_distribution(contributors_distribution)
+
+    # calculate to which percentage each group influences the total grade
+    contributions_percentage = settings.CONTRIBUTIONS_PERCENTAGE if contributors_average is not None else 0
+    course_grade_questions_percentage = settings.COURSE_GRADE_QUESTIONS_PERCENTAGE if course_grade_questions_average is not None else 0
+    course_non_grade_questions_percentage = 1 - settings.CONTRIBUTIONS_PERCENTAGE - settings.COURSE_GRADE_QUESTIONS_PERCENTAGE if course_non_grade_questions_average is not None else 0
+    if contributors_average or course_grade_questions_average or course_non_grade_questions_average:
+        scale = 1 / sum([contributions_percentage, course_grade_questions_percentage, course_non_grade_questions_percentage])
+        contributions_percentage *= scale
+        course_grade_questions_percentage *= scale
+        course_non_grade_questions_percentage *= scale
+
+    # calculate average grade
+    average_grade = contributions_percentage * (contributors_average or 0) + \
+                    course_grade_questions_percentage * (course_grade_questions_average or 0) + \
+                    course_non_grade_questions_percentage * (course_non_grade_questions_average or 0)
+
+    # calculate average distribution
+    distributions = [
+        (contributors_distribution, contributions_percentage),
+        (course_grade_questions_distribution, course_grade_questions_percentage),
+        (course_non_grade_questions_distribution, course_non_grade_questions_percentage),
+    ]
+    average_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+    for distribution, percentage in distributions:
+        for answer, value in distribution.items():
+            average_distribution[answer] += percentage * value
```

Comment:
...all of this is computing a weighted average. how about we use a function called weighted_average like this:
```
def weighted_average_distribution(distributions, weights):
    sum_weights = sum(weights)
    scaled_weights = [weight / sum_weights for weight in weights]
    average_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
    for distribution, weight in zip(distributions, scaled_weights):
        average_distribution = sum_distributions(average_distribution, multipy_distribution(distribution, weight))
    return average_distribution 
```
or don't scale the weights but normalize at the end, i'm not sure.

for consistency, i would also use this weighted average function instead of the mix (which would require calculating a second, kinda redundant, weight, but still), and also create such a function for floats instead of distributions.

the reason for this is that for once, the basic math stuff gets moved out of this function, and also that `mix` is kind of a special case of the weighted average of three values here. and here the averaging is performed manually while `mix` is a separate function, so why not create a function for the general case and use that everywhere...